File: 2015/12_JSAbacusFramework_io/Day12.js
// Day 12: JSAbacusFramework.io
// https://adventofcode.com/2015/day/12

const utils = require('../../utils');

const day = 12;
const input = utils.readInputByDay(day);
const inputTest = utils.readInputByDay(day, '_test');

function part1(input) {
  const nums = (input.match(/-?\d+/g) ?? []).map(Number);
  return nums.reduce((sum, num) => sum + num, 0);
}

function part2(input) {
  const obj = JSON.parse(input);
  return sumExcludeRed(obj);

  function sumExcludeRed(obj) {
    if (Array.isArray(obj)) {
      return obj.reduce((acc, val) => acc + sumExcludeRed(val), 0);
    }

    if (typeof obj === 'object') {
      if (Object.values(obj).includes('red')) {
        return 0;
      }
      return Object.values(obj).reduce((acc, val) => acc + sumExcludeRed(val), 0);
    }

    if (typeof obj === 'number') {
      return obj;
    }

    return 0;
  }
}

console.log('Part 1:', part1(input)); // 18
console.time('Part 1');
console.log('Part 1:', part1(inputTest)); // 191164
console.timeEnd('Part 1');

console.log('Part 2:', part2(input)); // 18
console.time('Part 2');
console.log('Part 2:', part2(inputTest)); // 87842
console.timeEnd('Part 2');

File: 2015/12_JSAbacusFramework_io/Day12.clj
; Day 12: JSAbacusFramework.io
; https://adventofcode.com/2015/day/12

(ns day12
  (:require [clojure.data.json :as json]
            [clojure.string :as str]))

(defn sum-numbers [input]
  (->> (re-seq #"-?\d+" input)
       (map #(Integer/parseInt %))
       (reduce +)))

(defn sum-json-numbers
  ([data] (sum-json-numbers data false))
  ([data skip-red]
   (cond
     (number? data) data
     (vector? data) (reduce + (map #(sum-json-numbers % skip-red) data))
     (map? data) (if (and skip-red (some #(= % "red") (vals data)))
                   0
                   (reduce + (map #(sum-json-numbers % skip-red) (vals data))))
     :else 0)))

(defn part1 [input]
  (sum-numbers input))

(defn part2 [input]
  (sum-json-numbers (json/read-str input) true))

(let [input (slurp "Day12.txt")
      input-test (slurp "Day12_test.txt")]
  (println "Part 1:" (part1 input))
  (println "Part 1:" (part1 input-test))

  (println "Part 2:" (part1 input))
  (println "Part 2:" (part2 input-test)))